File: Breeze.Client/Scripts/IBlade/dataType.js

define(["core"],
function (core) {
    /**
    @module breeze
    **/

    var Enum = core.Enum;
    /**
    DataType is an 'Enum' containing all of the supported data types.

    @class DataType
    @static
    **/

    /**
    The default value of this DataType.
    @property defaultValue {any}
    **/

    /**
    Whether this is a 'numeric' DataType. 
    @property isNumeric {Boolean}
    **/

    var dataTypeMethods = {
        // default
    };

    var coerceToString = function (source, sourceTypeName) {
        return (source == null) ? source : source.toString();
    };

    var coerceToInt = function (source, sourceTypeName) {
        if (sourceTypeName === "string") {
            var val = parseInt(source, 10);
            return isNaN(val) ? source : val;
        } else if (sourceTypeName === "number") {
            return Math.round(source);
        }
        // do we want to coerce floats -> ints
        return source;
    };

    var coerceToFloat = function (source, sourceTypeName) {
        if (sourceTypeName === "string") {
            var val = parseFloat(source);
            return isNaN(val) ? source : val;
        }
        return source;
    };

    var coerceToDate = function (source, sourceTypeName) {
        if (sourceTypeName === "string") {
            var val = new Date(Date.parse(source));
            return core.isDate(val) ? val : source;
        } else if (sourceTypeName === "number") {
            var val = new Date(source);
            return core.isDate(val) ? val : source;
        }
        return source;
    };

    var coerceToBool = function (source, sourceTypeName) {
        if (sourceTypeName === "string") {
            var src = source.trim().toLowerCase();
            if (src === 'false') {
                return false;
            } else if (src === "true") {
                return true;
            } else {
                return source;
            }
        } 
        return source;
    };

    var DataType = new Enum("DataType", dataTypeMethods);
    
    
    /**
    @property String {DataType}
    @final
    @static
    **/
    DataType.String = DataType.addSymbol({ defaultValue: "", parse: coerceToString });
    /**
    @property Int64 {DataType}
    @final
    @static
    **/
    DataType.Int64 = DataType.addSymbol({ defaultValue: 0, isNumeric: true, isInteger: true, parse: coerceToInt });
    /**
    @property Int32 {DataType}
    @final
    @static
    **/
    DataType.Int32 = DataType.addSymbol({ defaultValue: 0, isNumeric: true, isInteger: true, parse: coerceToInt });
    /**
    @property Int16 {DataType}
    @final
    @static
    **/
    DataType.Int16 = DataType.addSymbol({ defaultValue: 0, isNumeric: true, isInteger: true, parse: coerceToInt });
    /**
    @property Decimal {DataType}
    @final
    @static
    **/
    DataType.Decimal = DataType.addSymbol({ defaultValue: 0, isNumeric: true, parse: coerceToFloat });
    /**
    @property Double {DataType}
    @final
    @static
    **/
    DataType.Double = DataType.addSymbol({ defaultValue: 0, isNumeric: true, parse: coerceToFloat });
    /**
    @property Single {DataType}
    @final
    @static
    **/
    DataType.Single = DataType.addSymbol({ defaultValue: 0, isNumeric: true, parse: coerceToFloat });
    /**
    @property DateTime {DataType}
    @final
    @static
    **/
    DataType.DateTime = DataType.addSymbol({ defaultValue: new Date(1900, 0, 1), parse: coerceToDate });
    /**
    @property Time {DataType}
    @final
    @static
    **/
    DataType.Time = DataType.addSymbol({ defaultValue: "PT0S" });
    /**
    @property Boolean {DataType}
    @final
    @static
    **/
    DataType.Boolean = DataType.addSymbol({ defaultValue: false, parse: coerceToBool });
    /**
    @property Guid {DataType}
    @final
    @static
    **/
    DataType.Guid = DataType.addSymbol({ defaultValue: "00000000-0000-0000-0000-000000000000" });
    /**
    @property Byte {DataType}
    @final
    @static
    **/
    DataType.Byte = DataType.addSymbol({ defaultValue: 0 });
    /**
    @property Binary {DataType}
    @final
    @static
    **/
    DataType.Binary = DataType.addSymbol({ defaultValue: null });
    /**
    @property Undefined {DataType}
    @final
    @static
    **/
    DataType.Undefined = DataType.addSymbol({ defaultValue: undefined });
    DataType.seal();

    /**
    Returns the DataType for a specified EDM type name.
    @method fromEdmDataType
    @static
    @param typeName {String}
    @return {DataType} A DataType.
    **/
    DataType.fromEdmDataType = function (typeName) {
        var dt = null;
        var parts = typeName.split(".");
        if (parts.length > 1) {
            var simpleName = parts[1];
            if (simpleName === "image") {
                // hack
                dt = DataType.Byte;
            } else if (parts.length == 2) {
                dt = DataType.fromName(simpleName);
                if (!dt) {
                    if (simpleName === "DateTimeOffset") {
                        dt = DataType.DateTime;
                    } else {
                        dt = DataType.Undefined;
                    }
                }
            } else {
                // enum
                // dt = DataType.Int32;
                dt = DataType.String;
            }
        }

        return dt;
    };

    DataType.fromValue = function(val) {
        if (core.isDate(val)) return DataType.DateTime;
        switch (typeof val) {
            case "string":
                if (core.isGuid(val)) return DataType.Guid;
                else if (core.isDuration(val)) return DataType.Time;
                return DataType.String;
            case "boolean":
                return DataType.Boolean;
            case "number":
                return DataType.Int32;
        }
        return DataType.Undefined;
    };
   
    var _localTimeRegex = /.\d{3}$/;

    DataType.parseDateAsUTC = function (source) {
        if (typeof source === 'string') {
            // convert to UTC string if no time zone specifier.
            var isLocalTime = _localTimeRegex.test(source);
            source = isLocalTime ? source + 'Z' : source;
        }
        source = new Date(Date.parse(source));
        return source;
    };

    // NOT YET NEEDED --------------------------------------------------
    // var _utcOffsetMs = (new Date()).getTimezoneOffset() * 60000;
    
    //DataType.parseDateAsLocal = function (source) {
    //    var dt = DataType.parseDatesAsUTC(source);
    //    if (core.isDate(dt)) {
    //        dt = new Date(dt.getTime() + _utcOffsetMs);
    //    }
    //    return dt;
    //};
    // -----------------------------------------------------------------

    DataType.parseDateFromServer = DataType.parseDateAsUTC;

    return DataType;

});

File: Breeze.NetClient/DataType.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml;

namespace Breeze.Metadata {
  
  [Flags]
  public enum DataTypeInfo {
    None = 0,
    IsNumeric = 1,
    IsInteger = 2,
    IsDate = 4
  }

  public class NumericDataType : DataType  {
    public NumericDataType(Type clrType, String fmtSuffix, bool isInteger) : base(clrType) {
      DefaultValue = Convert.ChangeType(0, clrType);
      FmtOData = (Object source) => FmtNumber(source, clrType, fmtSuffix);
      GetNext = () => GetNextValue(clrType);
      DataTypeInfo = DataTypeInfo.IsNumeric | (isInteger ? DataTypeInfo.IsInteger : DataTypeInfo.None);
    }
  }

  public class DataType {
    public DataType(Type clrType) {
      Name = clrType.Name;
      ClrType = clrType;
      All.Add(this);
    }
    public String Name { get; internal set; }
    public Type ClrType { get; internal set; }
    public Object DefaultValue { get; internal set; }
    public virtual Object Parse(Object val) { 
      if (val == null) return null;
      return Convert.ChangeType(val, ClrType);
    }
    public Func<Object, String> FmtOData { get; internal set; }
    public Func<Object> GetNext { get; internal set;}
    public DataTypeInfo DataTypeInfo { get; internal set; }
    public static List<DataType> All = new List<DataType>();   
    
    public static DataType String = new DataType(typeof(String)) {
        DefaultValue = "",
        FmtOData =  FmtString,
        GetNext =  null
    };

    public static DataType Int64 = new NumericDataType(typeof(Int64), "L", true);

    public static DataType Int32 = new NumericDataType(typeof(Int32), "", true);

    public static DataType Int16 = new NumericDataType(typeof(Int16), "", true);

    public static DataType Byte  = new NumericDataType(typeof(Byte), "", true);

    public static DataType Decimal = new NumericDataType(typeof(Decimal), "m", false);

    public static DataType Double = new NumericDataType(typeof(Double), "d", false);

    public static DataType Single = new NumericDataType(typeof(Single), "f", false);

    public static DataType DateTime = new DataType(typeof(DateTime)) {
      DefaultValue = new DateTime(1900, 1, 1),
      FmtOData = FmtDateTime,
      GetNext = null,
      DataTypeInfo = DataTypeInfo.IsDate
    };

    public static DataType DateTimeOffset = new DataType(typeof(DateTimeOffset)) {
      DefaultValue = new DateTimeOffset(1900, 1, 1,0,0,0, new TimeSpan()),
      FmtOData = FmtDateTime,
      DataTypeInfo = DataTypeInfo.IsDate
    };

    public static DataType Time = new DataType(typeof(TimeSpan)) {
      Name = "Time",
      DefaultValue = new TimeSpan(0),
      FmtOData = FmtTime,
    };

    public static DataType Boolean = new DataType(typeof(Boolean)) {
      DefaultValue = false,
      FmtOData = FmtBoolean,
    };

    public static DataType Guid = new DataType(typeof(Guid)) {
      DefaultValue = System.Guid.Empty,
      FmtOData = FmtGuid,
      GetNext = () => GetNextValue(typeof(Guid))
    };

    public static DataType Binary = new DataType(typeof(Byte[])) {
      Name = "Binary",
      DefaultValue = null,
      FmtOData = FmtBinary,
    };

    public static DataType Undefined = new DataType(typeof(Object)) {
      DefaultValue = null,
      FmtOData = FmtUndefined,
    };

    public static DataType FromName(String typeName) {
      var dataType = All.FirstOrDefault(dt => dt.Name == typeName);
      return dataType != null ? dataType : DataType.Undefined;
    }

    public static DataType FromEdmType(String typeName) {
      DataType dt = DataType.Undefined;        
      var parts = typeName.Split('.');
      if (parts.Length > 1) {
          var simpleName = parts[1];
          if (simpleName == "image") {
              // hack
              dt = DataType.Byte;
          } else if (parts.Length == 2) {
            dt = DataType.FromName(simpleName);
          } else {
              // enum
              // dt = DataType.Int32;
              dt = DataType.String;
          }
      }
      return dt;
    }
   

    protected static Object GetNextValue(Type t) {
      // TODO: implement this for numbers, guids and dateTimes
      return null;
    }



    protected static String FmtString(Object val) {
      if (val == null) return null;
      return "'" + ((String)val).Replace("'", "''") + "'";
    }

    protected static String FmtNumber(Object val, Type clrType, String fmtSuffix) {
      if (val == null) return null;
      var tmp = Convert.ChangeType(val, clrType);
      return tmp.ToString() + fmtSuffix;
    }

    protected static String FmtDateTime(Object val) {
      if (val == null) return null;
      var date = (DateTime) Convert.ChangeType(val, typeof(DateTime));
      var tmp = date.ToString("s", System.Globalization.CultureInfo.InvariantCulture);
      return "datetime'" + tmp + "'";
    }

    protected static String FmtDateTimeOffset(Object val) {
      if (val == null) return null;
      var date = (DateTimeOffset)Convert.ChangeType(val, typeof(DateTimeOffset));
      var tmp = date.ToString("s", System.Globalization.CultureInfo.InvariantCulture);
      return "datetimeoffset'" + tmp + "'";
    }

    protected static String FmtTime(Object val) {
      if (val == null) return null;
      var timeSpan = (TimeSpan)val;
      var tmp = XmlConvert.ToString(timeSpan);
      return "time'" + tmp + "'";
    }

    protected static String FmtGuid(Object val) {
      if (val == null) return null;
      var guid = (Guid)val;
      var tmp = guid.ToString();
      return "guid'" + tmp + "'";
    }

    protected static String FmtBoolean(Object val) {
      if (val == null) return null;
      var tmp  = (Boolean)val;
      return  tmp ? "true" : "false";
    }

    protected static String FmtBinary(Object val) {
      if (val == null) return null;
      var tmp = (Byte[])val;
      return "binary'" + tmp + "'";
    }

    protected static String FmtUndefined(Object val) {
      if (val == null) return null;
      return val.ToString();
    }


  }
}
